src/vbk/p2p_sync.cpp

```diff
@@ -1,57 +1,145 @@
 #include "vbk/p2p_sync.hpp"
 
-#include <vbk/pop_service.hpp>
-
 #include "veriblock/entities/atv.hpp"
 #include "veriblock/entities/vbkblock.hpp"
 #include "veriblock/entities/vtb.hpp"
 
-#include "veriblock/mempool.hpp"
-
 namespace VeriBlock {
 namespace p2p {
 
-bool processPopData(CNode* pfrom, const std::string& strCommand, CDataStream& vRecv, int64_t nTimeReceived, const CChainParams& chainparams, CConnman* connman, BanMan* banman, const std::atomic<bool>& interruptMsgProc)
+std::map<NodeId, PopDataNodeState> mapPopDataNodeState;
+
+template <typename PopDataType>
+bool processGetPopData(CNode* node, CConnman* connman, CDataStream& vRecv, altintegration::MemPool& pop_mempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    AssertLockHeld(cs_main);
+    std::vector<std::vector<uint8_t>> requested_data;
+    vRecv >> requested_data;
+
+    if (requested_data.size() > MAX_POP_DATA_SENDING_AMOUNT) {
+        Misbehaving(node->GetId(), 20, strprintf("message getdata size() = %u", requested_data.size()));
+        return false;
+    }
+
+    auto& known_set = mapPopDataNodeState[node->GetId()].getSet<PopDataType>();
+
+    const CNetMsgMaker msgMaker(PROTOCOL_VERSION);
+    for (const auto& data_hash : requested_data) {
+        const PopDataType* data = pop_mempool.get<PopDataType>(data_hash);
+        known_set.insert(data_hash);
+        if (data != nullptr) {
+            connman->PushMessage(node, msgMaker.Make(PopDataType::name(), *data));
+        }
+    }
+
+    return true;
+}
+
+template <typename PopDataType>
+bool processOfferPopData(CNode* node, CConnman* connman, CDataStream& vRecv, altintegration::MemPool& pop_mempool) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    AssertLockHeld(cs_main);
+    LogPrint(BCLog::NET, "received offered pop data: %s, bytes size: %d\n", PopDataType::name(), vRecv.size());
+    std::vector<std::vector<uint8_t>> offered_data;
+    vRecv >> offered_data;
+
+    if (offered_data.size() > MAX_POP_DATA_SENDING_AMOUNT) {
+        Misbehaving(node->GetId(), 20, strprintf("message getdata size() = %u", offered_data.size()));
+        return false;
+    }
+
+    auto& known_set = mapPopDataNodeState[node->GetId()].getSet<PopDataType>();
+
+    std::vector<std::vector<uint8_t>> requested_data;
+    const CNetMsgMaker msgMaker(PROTOCOL_VERSION);
+    for (const auto& data_hash : offered_data) {
+        if (!pop_mempool.get<PopDataType>(data_hash)) {
+            requested_data.push_back(data_hash);
+            known_set.insert(data_hash);
+        }
+
+        if (requested_data.size() == MAX_POP_DATA_SENDING_AMOUNT) {
+            connman->PushMessage(node, msgMaker.Make(get_prefix + PopDataType::name(), requested_data));
+            requested_data.clear();
+        }
+    }
+
+    if (!requested_data.empty()) {
+        connman->PushMessage(node, msgMaker.Make(get_prefix + PopDataType::name(), requested_data));
+    }
```

Comment:
you can do PushMessage once, as requested_data in the loop will never be > MAX_POP_DATA_SENDING_AMOUNT

Reply:
fixed